kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/TransactionCommitLoopSuite.scala

```diff
@@ -0,0 +1,189 @@
+    import org.apache.spark.sql.functions.col
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialTxn = createWriteTxnBuilder(table)
+        .withSchema(engine, testSchema)
+        .build(engine)
+      commitTransaction(initialTxn, engine, emptyIterable()) // 000.json
+
+      val kernelTxn = createWriteTxnBuilder(table).withMaxRetries(5).build(engine)
+
+      // Create 001.json. This will make the engine throw a FileAlreadyExistsException when trying
+      // to write 001.json. The default committer will turn this into a
+      // CFE(isRetryable=true, isConflict=true).
+
+      spark.range(0, 10)
+        .select(col("id").cast("int"))
+        .write.format("delta").mode("append").save(tablePath)
+
+      val result = commitTransaction(kernelTxn, engine, emptyIterable())
+
+      assert(result.getVersion == 2)
+      assert(result.getTransactionReport.getTransactionMetrics.getNumCommitAttempts == 2)
+    }
+  }
+
+  test("Txn will attempt to re-commit *same* version on CFE(isRetryable=true, isConflict=false)") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialTxn = createWriteTxnBuilder(table).withSchema(engine, testSchema).build(engine)
+      commitTransaction(initialTxn, engine, emptyIterable()) // 000.json
+
+      var attemptCount = 0 // Will be incremented when actual writeJson attempt occurs
+      val attemptNumberToSucceedAt = 5
+      val attemptedFilePaths = scala.collection.mutable.Set[String]()
+
+      class CustomJsonHandler extends DefaultJsonHandler(fileIO) {
+        override def writeJsonFileAtomically(
+            filePath: String,
+            data: CloseableIterator[Row],
+            overwrite: Boolean): Unit = {
+          attemptCount += 1
+          attemptedFilePaths += filePath
+          if (attemptCount < attemptNumberToSucceedAt) {
+            // The default committer will turn this into a CFE(isRetryable=true, isConflict=true)
+            throw new java.io.IOException("Transient network error")
+          }
+          super.writeJsonFileAtomically(filePath, data, overwrite)
+        }
+      }
+
+      class CustomEngine extends DefaultEngine(fileIO) {
+        override def getJsonHandler: JsonHandler = new CustomJsonHandler()
+      }
+
+      val transientErrorEngine = new CustomEngine()
+      val txn = createWriteTxnBuilder(table).build(transientErrorEngine)
+      val result = commitTransaction(txn, transientErrorEngine, emptyIterable())
+
+      assert(result.getVersion == 1)
+      assert(attemptCount == attemptNumberToSucceedAt)
+      assert(attemptedFilePaths.size == 1) // we should only be attempting to write 001.json
+      assert(result.getTransactionReport.getTransactionMetrics.getNumCommitAttempts ==
+        attemptNumberToSucceedAt)
+    }
+  }
+
+  // TODO: Transaction will fail on CFE(isRetryable=false, isConflict=true/false). The default
```

Comment:
Is it worth checking that we *don't* retry for misc. runtime exceptions? wdyt?

Reply:
sg!